### PR TITLE
[bugFix] Support `masked_fill.Scalar`

### LINKED ISSUE #266: [Tags: v2.1] Error behaviorfor 'flag_gems.masked_fill(inp, mask, value)'
### Describe the bug

位置：
FlagGems/tests/test_unary_pointwise_ops.py::test_accuracy_masked_fill
FlagGems/src/flag_gems/ops/masked_fill.py

**现象1：**
在masked_fill.py中加打印， with flag_gems.use_gems():使用时不会出现打印，flag_gems.masked_fill(inp, mask, value)使用时才会走打印引处
**怀疑原因：**
算子没替换好

**现象2：**
flag_gems.masked_fill(inp, mask, value)使用时，‘line61: value = value.contiguous()’出错
**怀疑原因：**
value一般输入是int，不需要.contiguous()

### LINKED ISSUE #299: masked_fill op bug
### Describe the bug

1. Registration bug. `flag_gems.masked_fill` implementation is not correctly registered for `torch.masked_fill` op. The op names in aTen are `masked_fill_.Tensor`  and `masked_fill_.Scalar`.
2. `flag_gems.masked_fill` has some bugs: 
* Variables in the config mismatch the args in the triton kernel
"TypeError: dynamic_func() missing 2 required positional arguments: 'BLOCK_M' and 'BLOCK_N'"
* `value = value.contiguous()` raises error 
"AttributeError: 'int' object has no attribute 'contiguous"
* When inp is (), `N = inp.size(inp.ndim - 1)` raises error 
"IndexError: Dimension specified as -1 but tensor has no dimensions"

### PR DESCRIPTION
**Pull request by GwokHiujin**
### PR Category

Operator

### Type of Change

Bug Fix

### Description

Now can correctly:
- Process scalar-type value and single-number-type value
- Handle inp like `torch.tensor([])` and zero-dim inp

### Issue

- Resolves #266 
- Resolves #299 

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### CODE REVIEW

File: src/flag_gems/__init__.py [hunk 1]
```diff
@@ -147,6 +147,9 @@ def enable(lib=aten_lib):
     lib.impl("index_select", index_select, "CUDA")
     lib.impl("tile", tile, "CUDA")
     lib.impl("masked_fill", masked_fill, "CUDA")
+    lib.impl("masked_fill.Scalar", masked_fill, "CUDA")
+    lib.impl("masked_fill_", masked_fill_, "CUDA")
```
Comment (StrongSpoon) [inline src/flag_gems/__init__.py]: there are no interfaces named 'masked_fill' and 'masked_fill_' in native_functions.yaml. I think we should delete line 149 and line 151.

File: tests/test_reduction_ops.py [hunk 2]
```diff
@@ -595,3 +595,26 @@ def test_accuracy_masked_select(shape, dtype, threshold):
...
+
+
+@pytest.mark.masked_fill
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+@pytest.mark.parametrize("threshold, shape", THRESHOLD_SHAPE)
+@pytest.mark.parametrize(
+    "value",
+    [torch.tensor(1024, device="cuda"), torch.scalar_tensor(1024, device="cuda"), 1024],
+)
+def test_accuracy_masked_fill(shape, dtype, threshold, value):
```
Comment (StrongSpoon) [inline tests/test_reduction_ops.py]: test for masked_fill already exists in test_unary_ponitwise_ops.py. you could add a test case of inplace masked_fill_ in that file.

File: src/flag_gems/ops/masked_fill.py [hunk 3]
```diff
@@ -70,3 +88,33 @@ def masked_fill(inp, mask, value):
     )
     masked_fill_kernel[grid](inp, expand_mask.to(torch.int), value, out, M, N)
     return out
+
+
+def masked_fill_(inp, mask, value):
+    logging.debug("GEMS MASKED FILL")
+    assert (
+        (torch.is_tensor(value) and value.ndim == 0 and value.numel() == 1)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/masked_fill.py]: value.ndim == 0 and value.numel() == 1 overlap. I suggest preserving the former.

File: src/flag_gems/ops/masked_fill.py [hunk 4]
```diff
@@ -70,3 +88,33 @@ def masked_fill(inp, mask, value):
...
+    assert (
+        (torch.is_tensor(value) and value.ndim == 0 and value.numel() == 1)
+        or isinstance(value, int)
+        or isinstance(value, float)
+    ), "masked_fill_ only supports a 0-dimensional value tensor"
+    if torch.is_tensor(value):
+        # Value can be a tensor or a scalar
+        value = value.item()
+    inp_shape = tuple(inp.shape)
+    mask_shape = tuple(mask.shape)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/masked_fill.py]: is it necessary to convert shape to tuple?

File: src/flag_gems/ops/masked_fill.py [hunk 5]
```diff
@@ -58,15 +77,50 @@ def masked_fill(inp, mask, value):
...
+    inp = inp.contiguous()
+    mask = mask.contiguous()
+    expand_mask = mask.expand(inp.shape)
+
+    index = (inp.ndim - 1) if inp.ndim > 0 else 0
+    N = inp.size(index)
+    if N == 0:
+        return inp
+    M = inp.numel() // N
+    grid = lambda meta: (
```
Comment (StrongSpoon) [inline src/flag_gems/ops/masked_fill.py]: maybe we could regard inp and mask tensor as 1-dim vector?

File: src/flag_gems/__init__.py (hunk 1 above)
Comment (GwokHiujin) [reply src/flag_gems/__init__.py]: Modified them to *.Tensor

File: tests/test_reduction_ops.py (hunk 2 above)
Comment (GwokHiujin) [reply tests/test_reduction_ops.py]: Done.

File: src/flag_gems/ops/masked_fill.py (hunk 3 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/masked_fill.py]: Done.

File: src/flag_gems/ops/masked_fill.py (hunk 4 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/masked_fill.py]: Can be simplified. Now I've deleted them :)

File: src/flag_gems/ops/masked_fill.py (hunk 5 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/masked_fill.py:135]: I tested some edge cases in pyTorch and found that it indeed can handle 0-dim input, maybe keep up with it is better?

Comment (GwokHiujin) [reply src/flag_gems/ops/masked_fill.py:131]: Considering torch's behavior, I think it may introduce unnecessary operations.

Comment (GwokHiujin) [reply src/flag_gems/ops/masked_fill.py]: Looks good.

**Review by StrongSpoon — APPROVED** [review]
lgtm

**Pull request merged**